### ISSUE TITLE
error when overriding optional field with mandatory field on dataclass, and false negative when attempting to instantiate the subclass without the mandatory field

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#835]
Code sample in [basedpyright playground](https://basedpyright.com/?pyrightVersion=1.17.3&pythonVersion=3.12&typeCheckingMode=all&code=GYJw9gtgBAJghgFzgYwDZwM4YKYagSwgAcwQFZEV0sBYAKHvoAF4k1MMAKUMAL2wB2AXgAqIAK7YANFADWAdwD6YAagCeoidgCU9dligAxMGABc9KJahxTBAeQA%2BUAHIrsUIS7eMGdFpX0uHn5hMUkZBWVVDTCdPWo8ACE4EE5jMG1zOitrW3x7KABiKGwQcBAfekUPKGTU7SKoATASstIgA)

```python
from dataclasses import dataclass


@dataclass(frozen=True, kw_only=True)
class Foo:
    a: int | None = None


@dataclass(frozen=True, kw_only=True)
class Bar(Foo):
    a: int # error


_ = Bar() # no error
```

i think this is intentional because at runtime it lets you instantiate the class. but this looks more like a bug in dataclasses rather than an intentional decision that the type checker should blindly allow.

mypy shows an error on the instantiation of `Bar` instead, which i think is the correct approach

**Comment by DetachHead** [comment]
https://github.com/microsoft/pyright/issues/8766#issuecomment-2291565896
